src/lnp/session/noise/conduit.rs

```diff
@@ -0,0 +1,630 @@
+
+        let mut ciphertext = vec![
+            0u8;
+            TAGGED_MESSAGE_LENGTH_HEADER_SIZE
+                + length as usize
+                + chacha::TAG_SIZE
+        ];
+
+        chacha::encrypt(
+            &self.sending_key,
+            self.sending_nonce as u64,
+            &[0; 0],
+            &length_bytes,
+            &mut ciphertext[..TAGGED_MESSAGE_LENGTH_HEADER_SIZE],
+        )?;
+        self.increment_nonce();
+
+        &chacha::encrypt(
+            &self.sending_key,
+            self.sending_nonce as u64,
+            &[0; 0],
+            buffer,
+            &mut ciphertext[TAGGED_MESSAGE_LENGTH_HEADER_SIZE..],
+        )?;
+        self.increment_nonce();
+
+        Ok(ciphertext)
+    }
+
+    fn increment_nonce(&mut self) {
+        Conduit::increment_nonce(
+            &mut self.sending_nonce,
+            &mut self.sending_chaining_key,
+            &mut self.sending_key,
+        );
+    }
+}
+
+impl Encrypt for Encryptor {
+    fn encrypt(&mut self, buffer: impl Borrow<[u8]>) -> Vec<u8> {
+        match self.encrypt(buffer.borrow()) {
+            Ok(values) => return values,
+            Err(_) => return Vec::new(),
+        }
+    }
+}
+
+pub struct Decryptor {
+    receiving_key: SymmetricKey,
+    receiving_chaining_key: SymmetricKey,
+    receiving_nonce: u32,
+
+    pending_message_length: Option<usize>,
+    read_buffer: Option<Vec<u8>>,
+    poisoned: bool, /* signal an error has occurred so None is returned on
+                     * iteration after failure */
+}
+
+impl Decryptor {
+    pub fn read(&mut self, data: &[u8]) {
+        let read_buffer = self.read_buffer.get_or_insert(Vec::new());
+        read_buffer.extend_from_slice(data);
+    }
+
+    /// Decrypt a single message. If data containing more than one message has
+    /// been received, only the first message will be returned, and the rest
+    /// stored in the internal buffer. If a message pending in the buffer
+    /// still hasn't been decrypted, that message will be returned in lieu
+    /// of anything new, even if new data is provided.
+    pub fn decrypt_single_message(
+        &mut self,
+        new_data: Option<&[u8]>,
+    ) -> Result<Option<Vec<u8>>, Error> {
+        let mut read_buffer = if let Some(buffer) = self.read_buffer.take() {
+            buffer
+        } else {
+            Vec::new()
+        };
+
+        if let Some(data) = new_data {
+            read_buffer.extend_from_slice(data);
+        }
+
+        if read_buffer.len() > LN_MAX_MSG_LEN + 16 {
+            panic!(
+                "Attempted to decrypt message longer than 65535 + 16 bytes!"
+            );
+        }
+
+        let (current_message, offset) = self.decrypt(&read_buffer[..])?;
+        read_buffer.drain(..offset); // drain the read buffer
+        self.read_buffer = Some(read_buffer); // assign the new value to the built-in buffer
+        Ok(current_message)
+    }
+
+    fn decrypt(
```

Comment:
Not sure whether it's nice idea to have the same function name in the impl of the type as in the trait. Probably `decrypt_raw` here will work better

Reply:
Noted.